internal/pkg/bulk/retry.go

```diff
@@ -0,0 +1,61 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package bulk
+
+import (
+	"regexp"
+	"time"
+
+	"github.com/rs/zerolog/log"
+)
+
+const maxRequestDelay = 5 * time.Minute
+
+func RetryOnUnavailable(e error) bool {
+	match, err := regexp.MatchString("connect: connection refused$|net/http: timeout awaiting response headers$", e.Error())
+	if err != nil {
+		log.Warn().Err(err).Msg("Error occured while parsing error message")
+		return false
+	}
+	return match
+}
+
+type retryActionT func() respT
+
+func RetryWithBackoff(action retryActionT, requestRetryLimit int, shouldRetry RequestRetryPredicateT, requestRetryDelayIncrement time.Duration) respT {
```

Comment:
Do we need a requestRetryLimit? What happens after the limit is reached?

Reply:
As mentioned below, we are going to solve this problem higher up for now.  In the above code, when retry limit it hit the function will error out, and you would still have to handle it upstream.